File: app-laravel/src/Product/Application/Commands/UpdateStock/UpdateStockHandler.php
<?php

declare(strict_types=1);

namespace Src\Product\Application\Commands\UpdateStock;

use Src\Shared\Application\Bus\CommandHandler;
use Src\Shared\Domain\Exceptions\DomainException;
use Src\Product\Domain\Repositories\ProductRepositoryInterface;
use Src\Product\Domain\ValueObjects\ProductId;

/**
 * UpdateStockHandler - Stok yeniləmə əmrini icra edən handler.
 *
 * Bu handler:
 * 1. Məhsulu repository-dən tapır.
 * 2. Stoku artırır və ya azaldır (type-a görə).
 * 3. Dəyişiklikləri saxlayır.
 *
 * Diqqət: Stok azaldıqda Product avtomatik olaraq StockDecreasedEvent
 * və lazım gəldikdə LowStockIntegrationEvent qeydə alır.
 * Handler bu event-lər barədə narahat olmur - bu, Domain-in işidir.
 */
final class UpdateStockHandler implements CommandHandler
{
    public function __construct(
        private readonly ProductRepositoryInterface $repository,
    ) {
    }

    /**
     * @param UpdateStockCommand $command Stok yeniləmə əmri
     * @throws DomainException Əgər məhsul tapılmadıqda və ya stok kifayət deyilsə
     */
    public function handle(UpdateStockCommand $command): void
    {
        // 1. Məhsulu tapırıq
        $product = $this->repository->findById(
            ProductId::fromString($command->productId)
        );

        if ($product === null) {
            throw new DomainException(
                "Məhsul tapılmadı. ID: {$command->productId}"
            );
        }

        // 2. Əməliyyat növünə görə stoku yeniləyirik
        match ($command->type) {
            'increase' => $product->increaseStock($command->amount),
            'decrease' => $product->decreaseStock($command->amount),
            default => throw new DomainException(
                "Yanlış əməliyyat növü: {$command->type}. "
                . "'increase' və ya 'decrease' olmalıdır."
            ),
        };

        // 3. Dəyişiklikləri saxlayırıq
        $this->repository->save($product);
    }
}

File: app/spring/src/main/java/az/ecommerce/product/application/command/updatestock/UpdateStockHandler.java
package az.ecommerce.product.application.command.updatestock;

import az.ecommerce.product.domain.Product;
import az.ecommerce.product.domain.repository.ProductRepository;
import az.ecommerce.product.domain.valueobject.ProductId;
import az.ecommerce.shared.application.bus.CommandHandler;
import az.ecommerce.shared.domain.exception.EntityNotFoundException;
import az.ecommerce.shared.infrastructure.bus.EventDispatcher;
import org.springframework.stereotype.Service;

@Service
public class UpdateStockHandler implements CommandHandler<UpdateStockCommand, Void> {

    private final ProductRepository repository;
    private final EventDispatcher eventDispatcher;

    public UpdateStockHandler(ProductRepository repository, EventDispatcher eventDispatcher) {
        this.repository = repository;
        this.eventDispatcher = eventDispatcher;
    }

    @Override
    public Void handle(UpdateStockCommand cmd) {
        Product product = repository.findById(new ProductId(cmd.productId()))
                .orElseThrow(() -> new EntityNotFoundException("Product", cmd.productId().toString()));

        if ("increase".equals(cmd.type())) {
            product.increaseStock(cmd.amount());
        } else {
            product.decreaseStock(cmd.amount());
        }

        repository.save(product);
        eventDispatcher.dispatchAll(product);
        return null;
    }
}
